packages/wrangler/src/deploy/config-diffs.ts

```diff
@@ -0,0 +1,296 @@
+	};
+}
+
+/**
+ * Given a diff object, representing the diff of config files, it computes whether such diff
+ * object only represents at most additions (so no removal nor modifications).
+ *
+ * For example the diff:
+ *  ```
+ *   - "name": "my-worker"
+ *   + "name": "my-worker",
+ *   + "vars": {
+ *   +   MY_VAR: "my variable",
+ *   + },
+ *  ```
+ * only contains additions.
+ *
+ * While
+ *  ```
+ *    "name": "my-worker",
+ *    "vars": {
+ *   -   MY_VAR: "my variable",
+ *   +   MY_VAR: "my modified variable",
+ *    },
+ *	 - compatibility_date: "2025-07-08",
+ *  ```
+ * contains also a modification and a removal
+ *
+ * @param diff The diff object to analyze
+ * @returns true if there are only additions or no diffs at all, false otherwise
+ */
+function configDiffOnlyHasAdditionsIfAny(diff: Diff): boolean {
+	const diffLines = `${diff}`.split("\n");
+	let currentRemovalIdx = 0;
+	while (currentRemovalIdx !== -1) {
+		const nextRemovalIdx = diffLines.findIndex((line, idx) => {
+			// We only consider values after the currentRemovalIdx (which is practically our starting index)
+			if (idx < currentRemovalIdx) {
+				return false;
+			}
+
+			const withoutLeadingSpaces = line.replace(/^\s*/, "");
+			return withoutLeadingSpaces.startsWith(red("-"));
+		});
+		if (nextRemovalIdx === -1) {
+			// We've looked for all the removals none were actually modifications
+			// so we return true, at most the changes in the diff are additions
+			return true;
+		}
+		currentRemovalIdx = nextRemovalIdx;
+		const lineAtIdx = diffLines[currentRemovalIdx];
+		const nextLine = diffLines[currentRemovalIdx + 1] ?? "";
+		const lineAtIdxButAdditionAndWithComma = `${lineAtIdx.replace(red("-"), green("+"))},`;
+		// onlyACommaWasAdded indicates that only a single comma was added, so
+		// for example, `lineAtIdx` is `'- "field": "test"'` and `nextLine` is `'+ "field': "test",`
+		const onlyACommaWasAdded = nextLine === lineAtIdxButAdditionAndWithComma;
+		if (!onlyACommaWasAdded) {
+			// if there is a removal and the change wasn't the simple addition of a comma
+			// then we've found a real removal/modification, so let's return false
+			return false;
+		}
+		// otherwise this wasn't a real removal/modification so we continue looking for one
+		currentRemovalIdx++;
+	}
+	// we didn't find any real removal/modification
+	return true;
+}
+
+/**
+ * Normalizes a local config object into an object that can be compared to the remote representation of
+ * a Worker's config. This includes resolving some values and setting the default values to some others.
+ *
+ * @param localConfig The local config to normalize
+ * @param localResolvedConfig The resolved local config (used to resolve some values on the raw config)
+ * @returns The local config normalized, ready to be compared to a remote one
+ */
+function normalizeLocalConfigAsRemote(
+	localConfig: RawConfig,
+	localResolvedConfig: Config
+): RawConfig {
```

Comment:
Why does this need a raw config and a resolved one? Can't it just take a `Config` object and transform that?

Reply:
As mentioned in my other comment (https://github.com/cloudflare/workers-sdk/pull/10314#discussion_r2269623289) I think we need to deal with the raw config so that we can then show a valid diff to users (without extra stuff).

The resolved config gets passed here so that we can resolve the `main` field so that that can be compared with the one the remote config has (and I passed the full `Config` object just in case in the future other resolved fields might be needed).

I am also happy to just remove the `main` field in the check instead (since that doesn't seem super valuable in this local/remote comparison anyways).

Reply:
Yeah, let's remove the main field from comparison. That will almost certainly be different in most cases

Reply:
Removed 🙂